Bad error message when trying to assign to a field through raw pointer

If I try to assign to a value named read through a raw pointer, I get a particularly bad error message
```rs
struct Foo { read: i32 }

fn blah(x: *mut Foo) {
    x.read = 4;
}
```

```
error[E0615]: attempted to take value of method `read` on type `*mut Foo`
 --> src/lib.rs:4:7
  |
4 |     x.read = 4;
  |       ^^^^ method, not a field
  |
  = help: methods are immutable and cannot be assigned to

For more information about this error, try `rustc --explain E0615`.
error: could not compile `playground` due to previous error
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a491cb65ab0afb0f69c80d92b91d7f5

I think this happens because it finds `core::ptr::read`'s method version on `*mut T`. It would be nice if we suggested[^1] something like `(*x).read` instead.

[^1]: I guess theres a bit of trickery here since... well, technically for non-copy thats being initialized for the first time, the right thing would be `addr_of_mut!((*x).read).write(val)`... So maybe a fix shouldnt be offered, just a better diagnostic.

If the field isn't "shadowed" by a method name, it says something like
```
error[E0609]: no field `red` on type `*mut Foo`
 --> src/lib.rs:4:7
  |
4 |     x.red = 4;
  |     --^^^
  |     |
  |     help: `x` is a raw pointer; try dereferencing it: `(*x).red`
  ```
  which is probbaly just what it should say always.

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -46,6 +46,7 @@ use rustc_span::hygiene::DesugaringKind;
 use rustc_span::lev_distance::find_best_match_for_name;
 use rustc_span::source_map::Span;
 use rustc_span::symbol::{kw, sym, Ident, Symbol};
+use rustc_span::{BytePos, Pos};
 use rustc_trait_selection::traits::{self, ObligationCauseCode};
 
 impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
@@ -2063,7 +2064,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Some(span),
             );
         } else {
-            err.help("methods are immutable and cannot be assigned to");
+            let mut found = false;
+
+            if let ty::RawPtr(ty_and_mut) = expr_t.kind() {
+                if let ty::Adt(adt_def, _) = ty_and_mut.ty.kind() {
+                    if adt_def.variants.len() == 1
+                        && adt_def
+                            .variants
+                            .iter()
+                            .next()
+                            .unwrap()
+                            .fields
+                            .iter()
+                            .any(|f| f.ident == field)
+                    {
+                        if let Some(dot_loc) = expr_snippet.rfind('.') {
+                            found = true;
+                            err.span_suggestion(
+                                expr.span.with_hi(expr.span.lo() + BytePos::from_usize(dot_loc)),
+                                "to access the field, dereference first",
+                                format!("(*{})", &expr_snippet[0..dot_loc]),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                    }
+                }
+            }
+
+            if !found {
+                err.help("methods are immutable and cannot be assigned to");
+            }
         }
 
         err.emit();

--- a/src/test/ui/typeck/issue-91210-ptr-method.fixed
+++ b/src/test/ui/typeck/issue-91210-ptr-method.fixed
@@ -1,0 +1,15 @@
+// Regression test for issue #91210.
+
+// run-rustfix
+
+#![allow(unused)]
+
+struct Foo { read: i32 }
+
+unsafe fn blah(x: *mut Foo) {
+    (*x).read = 4;
+    //~^ ERROR: attempted to take value of method
+    //~| HELP: to access the field, dereference first
+}
+
+fn main() {}

--- a/src/test/ui/typeck/issue-91210-ptr-method.rs
+++ b/src/test/ui/typeck/issue-91210-ptr-method.rs
@@ -1,0 +1,15 @@
+// Regression test for issue #91210.
+
+// run-rustfix
+
+#![allow(unused)]
+
+struct Foo { read: i32 }
+
+unsafe fn blah(x: *mut Foo) {
+    x.read = 4;
+    //~^ ERROR: attempted to take value of method
+    //~| HELP: to access the field, dereference first
+}
+
+fn main() {}

--- a/src/test/ui/typeck/issue-91210-ptr-method.stderr
+++ b/src/test/ui/typeck/issue-91210-ptr-method.stderr
@@ -1,0 +1,11 @@
+error[E0615]: attempted to take value of method `read` on type `*mut Foo`
+  --> $DIR/issue-91210-ptr-method.rs:10:7
+   |
+LL |     x.read = 4;
+   |     - ^^^^ method, not a field
+   |     |
+   |     help: to access the field, dereference first: `(*x)`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0615`.

